scripts/generate_error_code.hs

```diff
@@ -0,0 +1,280 @@
+
+// NOTE: This file should not be modified directly.
+// Use `stack scripts/generate_error_code.hs` instead.
+
+\#if ERRORS_CODE_MLIGO
+\#else
+\#define ERRORS_CODE_MLIGO
+
+\#include "types.mligo"
+
+#{errorsToLigoText "Invalid Input Error Codes" invalidInputErrors}
+
+#{errorsToLigoText "Contract Configuration Error Codes" invalidConfigErrors}
+
+#{errorsToLigoText "Internal Error Codes" internalErrors}
+
+\#endif
+|]
+
+---------------------------------------------------------------------------------
+-- Markdown
+---------------------------------------------------------------------------------
+
+errorItemToMdText :: ErrorItem -> Text
+errorItemToMdText ErrorItem{..} =
+  "| " <> (show @Text eiCode) <> " | `" <> eiLabel <> "` | " <> eiDesc <> " |"
+
+errorsToMdText :: Text -> [ErrorItem] -> Text
+errorsToMdText title errors = if (length errors /= 0) then
+  [i| \#\#\#\# #{title}
+
+| Error Code       | Error Label      | Description                                           |
+|------------------|------------------|-------------------------------------------------------|
+#{unlines $ errorItemToMdText <$> errors}
+|] else ""
+
+mdErrorsTemplate :: Text
+mdErrorsTemplate =
+  [i|<!--
+- SPDX-FileCopyrightText: 2021 TQ Tezos
+
+- SPDX-License-Identifier: LicenseRef-MIT-TQ
+-->
+
+<!--
+NOTE: This file should not be modified directly.
+Use `stack scripts/generate_error_code.hs` instead.
+-->
+
+\#\# Error Codes
+
+Here is a summary of all the error codes thrown by the contract.
+(The list of errors may be inaccurate and incomplete, it will be updated during the implementation.)
+
+
+#{errorsToMdText "Invalid Input Error Codes" invalidInputErrors}
+
+#{errorsToMdText "Contract Configuration Error Codes" invalidConfigErrors}
+
+#{errorsToMdText "Internal Error Codes" internalErrors}
+
+|]
+
+---------------------------------------------------------------------------------
+-- Haskell
+---------------------------------------------------------------------------------
+
+haskellFnName :: ErrorItem -> Text
+haskellFnName = toCamel . eiLabel
+
+errorsToHaskellText :: Text -> [ErrorItem] -> Text
+errorsToHaskellText title errors = if (length errors /= 0) then
+  [i|----------------------------------------------------------------------------
+-- #{title}
+----------------------------------------------------------------------------
+#{unlines $ errorItemToHaskellText <$> errors}
+
+|] else ""
+
+errorItemToHaskellText :: ErrorItem -> Text
+errorItemToHaskellText err@ErrorItem{eiDesc, eiCode} =
+  [i|
+-- | #{eiDesc}
+#{fnName} :: Natural
+#{fnName} = #{eiCode}|]
+  where
+    fnName = haskellFnName err
+
+errorItemToTzipError :: ErrorItem -> Text
+errorItemToTzipError err@ErrorItem{eiDesc, eiCode} =
+  [i|
+  EStatic $ StaticError
+    { seError = ExpressionInt $ toInteger #{fnName}
+    , seExpansion = ExpressionString "#{eiDesc}"
+    , seLanguages = ["en"]
+    }|]
+  where
+    fnName = haskellFnName err
```

Comment:
@rinn7e does this suit all errors tho?

AFAIU the `error` field here should be the Michelson representations of exactly the error that will be thrown with `FAILWITH`, however here we have also errors that return not just a `nat` but also a `nat * unit`, `nat * string` and more, so this doesn't look correct to me, am I wrong?

Reply:
@pasqu4le  That make sense, I missed this.

I've added a condition in case of a pair:
```
errorItemToTzipError :: ErrorItem -> Text
errorItemToTzipError err@ErrorItem{eiDesc, eiCode, eiArg} =
  [i|
  EStatic $ StaticError
    { seError = #{error}
    , seExpansion = ExpressionString "#{eiDesc}"
    , seLanguages = ["en"]
    }|]
  where
    fnName = haskellFnName err
    (error :: Text) = case eiArg of
      NoArg ->
        [i|ExpressionInt $ toInteger #{fnName}|]
      ArgString val ->
        [i|ExpressionSeq [ ExpressionInt $ toInteger #{fnName}, ExpressionString "#{val}" ]|]
   ```
   
   Not sure if using `ExpressionSeq` for pair works here, hope to get your advice on this

Reply:
@rinn7e IIRC `ExpressionSeq` should work alright, although I think a `PrimExpr` for `Pair` woul be better.

However the thing is that we shouldn't really be doing that IMO, we have `toExpression` in `morley` to do this conversion.
This can be used on typed `Value`, which in turn can be obtained using `toVal` (from `IsoValue`).

Doing this instead should be the better option

Reply:
I forgot about `toExpression`, pushed an update to use that instead now.